third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,491 @@
+    /// (May not panic in the release version due to the use of debug_assert)
+    fn get_the_successor(&self, block: BlockId) -> BlockId {
+        let successors = self.cfg.successors(block);
+        debug_assert!(successors.len() == 1);
+        *successors.first().expect("successor block")
+    }
+
+    /// Gets the only successor of `block` which is not entry/exit block; panics if this is not the case.
+    /// (May not panic in the release version due to the use of debug_assert)
+    fn get_the_non_trivial_successor(&self, block: BlockId) -> BlockId {
+        let the_suc = self.get_the_successor(block);
+        debug_assert!(!self.is_trivial_block(the_suc));
+        the_suc
+    }
+
+    /// Adds an explicit jump to `to_block` to the end of `codes`
+    fn add_explicit_jump(&self, codes: &mut Vec<Bytecode>, to_block: BlockId) {
+        debug_assert!(!self.is_trivial_block(to_block));
+        let to_label = self.get_block_label(to_block).expect("label");
+        let attr_id = self
+            .code_blocks
+            .get(&to_block)
+            .expect("code block")
+            .first()
+            .expect("first instruction")
+            .get_attr_id();
+        codes.push(Bytecode::Jump(attr_id, to_label));
+    }
+
+    /// Returns the instructions of the block
+    fn block_instrs(&self, block_id: BlockId) -> &[Bytecode] {
+        self.code_blocks.get(&block_id).expect("block instructions")
+    }
+
+    /// Returns the label of the block or `None` if it doesn't start with a label
+    fn get_block_label(&self, block_id: BlockId) -> Option<Label> {
+        if let Bytecode::Label(_, label) = self
+            .block_instrs(block_id)
+            .first()
+            .expect("first instruction")
+        {
+            Some(*label)
+        } else {
+            None
+        }
+    }
+
+    /// Checks if the block is entry or exit block
+    fn is_trivial_block(&self, block: BlockId) -> bool {
+        block == self.cfg.entry_block() || block == self.cfg.exit_block()
+    }
+}
+
+struct RemoveEmptyBlock(ControlFlowGraphCodeGenerator);
+
+macro_rules! impl_deref {
+    ($struct_name:ident, $target_type:ty) => {
+        impl std::ops::Deref for $struct_name {
+            type Target = $target_type;
+
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+    };
+}
+
+macro_rules! impl_deref_mut {
+    ($struct_name:ident) => {
+        impl std::ops::DerefMut for $struct_name {
+            fn deref_mut(&mut self) -> &mut Self::Target {
+                &mut self.0
+            }
+        }
+    };
+}
+
+impl_deref!(RemoveEmptyBlock, ControlFlowGraphCodeGenerator);
+
+impl_deref_mut!(RemoveEmptyBlock);
+
+impl RemoveEmptyBlock {
+    /// Wrapper
+    pub fn new(generator: ControlFlowGraphCodeGenerator) -> Self {
+        Self(generator)
+    }
+
+    /// Eliminating branch/jump to jump
+    pub fn transform(&mut self) {
```

Comment:
Better to move the documentation of the transform here, IMO, instead of the module doc. Module doc can just say something briefly about this transform, whereas the exact semantics of the transform can be provided here.

Reply:
Done. Copied some of the detailed documentation of the transform method. I didn't delete the detailed explanation in the module doc because some of the passes (removes all jumps to the next instruction) are not done in a dedicated pass.

Reply:
Do not duplicate code documentation, that is, either document here or in the header. Duplicate documentation will very soon be inconsistent.

Reply:
done